### DISCUSSION TITLE
Bad Battery life on Watch since 'Nightly build 1st Oct 2020'

### DISCUSSION THREAD

**Discussion by oddy1968** [NightscoutFoundation/xDrip#3019] [Ideas]
Hi,
I think I have found a problem with Battery life on the Smartwatch since the '_Nightly build 1st Oct 2020_'.

I use the following setup for my little son:
Smartwatch is a TicWatch C2 with '_Force Wear_' enabled.
The watch is only for collecting BG, no incoming messages etc..
Phone is a Samsung S5 mini

After Oct. update the watch had 14% Battery after 12h, before around 60%.
Now I am back on '_Nightly build 4th Sep 2020_' and after 12h it shows me 60% again.

Can someone confirm that?

**Comment by tolot27** [comment]
That is strange. The difference between [2020.09.04 and 2020.10.01](https://github.com/NightscoutFoundation/xDrip/compare/2020.09.04...2020.10.01) is only the usage of [JodaTime](https://www.joda.org/joda-time) in the Android App. The wear app did not change.
It would be interesting to get some logs from the TicWatch.

**Comment by oddy1968** [comment]
I know. 
Only two apps are on the watch which updates. That are xdrip and wearOS.
At first I thought its a problem with the wear app but its a problem to downgrade.
Then I tried xdrip and after two weeks with 14% I had 60% again.

How can I get logs from the TicWatch?

**Comment by dbnsacc** [comment]
I can partially confirm this phenomenon. I'm using a TicWatch Pro 2020 and I've just reverted from the 2020.10.01 build back to 2020.09.04 on both phone and watch and the watch battery has been better over the last day since I did so.

I was actually investigating a strange kernel wakelock related to bluetooth (I think) on the phone which I couldn't trace back to an app and initially this seems to have alleviated itself also after going back but I intend to give it a few days before I'm sure that xDrip+ was the cause. The wakesource I was seeing was from alarmtimer and the screenshot below details the reason which I have been unable to find a cause for after searching.

Slim chance, but could this be related to the JodaTime change in 2020.10.01 @tolot27 or am I on the wrong track?
![shot-19112020_1732](https://user-images.githubusercontent.com/44261610/99794855-b5911a00-2b22-11eb-95f9-817562b88475.jpeg)

**Comment by tolot27** [comment]
Please add the extra logging tag joda-time-android:v. I don't think it gives us more information because logging is rarely used in the joda library.

I still have no idea why joda should affect the wear app. It is not packaged with it.

**Comment by oddy1968** [comment]
> Please add the extra logging tag joda-time-android:v. I don't think it gives us more information because logging is rarely used in the joda library.

Can you do this dbnsacc? thx

**Comment by radugreta** [comment]
Hello  I use p20 huawei and TicWatch S2 Glacier.   I am just a follower in xdrip. and the master have a A40 Samsung.
The watch connected to my phone and also to the masters phone, the battery was emply in less than 1 day.
We used e14bae1-2020.07.13.
Now I moved to 39e880d-2021.02.17 and it seems to be a much better battery life.
I put a screenshot with 2 screenshots from before and 1 after.

![Screenshot 2021-02-26 085641](https://user-images.githubusercontent.com/58292604/109266320-92a48f00-7810-11eb-9d0a-0b2b1fac99f6.png)

**Comment by Navid200** [comment]
How could xDrip affect the battery runtime of the watch?
Is this still a problem with latest Nightly builds?

**Comment by dbnsacc** [comment]
I'd need to update as I'm running 2021.08.18 at the moment but I think what it boils down to when looking at a debug dump in batteryhistorian is the amount of alarms and wakelocks xDrip+ makes with bluetooth. If I remove the complication from my watchface and force stop xDrip+ on the watch battery consumption usually reduces quite a bit until its reactivated. I can update both phone and watch to the latest nightly this week and get a new dump for battery historian if you like?

**Comment by Navid200** [comment]
@dbnsacc I don't have a grasp of the problem.  You may be able to help me with that.
Are you thinking we change xDrip so that it would send less frequent alarms to address this?
If yes, why would this have started with the October 1, 2020?
Regardless, if we reduce the number of updates, would that not reduce the update rate on watches?
Or, are you saying xDrip unnecessarily sends redundant info it shouldn't?

**Comment by dbnsacc** [comment]
I would have to install a version prior to 1st October 2020 to compare their behaviour again and get a dump as it's been a while now but I do remember battery consumption of xDrip seemed to increase from this build on my watch. It seems to me that there are potentially excessive wakelocks and alarms. I'll give a brief example from the last debug I took from my Ticwatch below. 

If the BG is updated every 5 minutes it would be reasonable to expect there to be bluetooth communication for xDrip every 5 minutes or so between watch and phone (which is the collector) which would be 12 times an hour (which does just about match up with 'watchlistener-onstart'). However my stats are showing other wakelocks and alarms at a rate of about 34 times an hour according to battery historian. There may well be good reasons for this that I don't know about but it seems (if we give some leeway from 12 times an hour to 17 times an hour) as if xDrip is firing alarms and holding wakelocks twice the amount it maybe could be?

Another app I use for heart rate which is set to every 30 minutes wakes up twice an hour and sends the results back to its companion phone app. It shows up as waking up around twice an hour as I'd expect in the stats.

![xdrip-1510_screenshot_001-27102021](https://user-images.githubusercontent.com/44261610/139116994-b64136b3-6934-43ec-8df3-36cefde841a2.jpg)
![xdrip-1510_screenshot_002-27102021](https://user-images.githubusercontent.com/44261610/139116999-fc8eb7b0-35c7-4ea3-88bd-7d65ea42fa92.jpg)
![xdrip-1510_screenshot_003-27102021](https://user-images.githubusercontent.com/44261610/139117000-756121a8-41e7-46a2-aeb0-54bb6570bf5e.jpg)
.

**Comment by Navid200** [comment]
I'm hoping that someone will move WearOS out of xDrip.  Then, WearOS can be better supported.
In the meantime, I am going to convert this to a discussion.
